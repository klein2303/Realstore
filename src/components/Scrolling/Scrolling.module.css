.clothesSections {
    margin-top: 20px;
    margin-bottom: 250px;
    padding: 25px;
}

.rows {
    display: flex;
    flex-direction: row;
    align-items: center;
    justify-content: center;
    flex-wrap: wrap;
    gap: 20px;
}

.message {
    justify-content: center;
    text-align: center;
}

@media (max-width: 800px) {
    .rows {
        flex-direction: column;
        justify-content: center;
        align-items: center;
    }
}
