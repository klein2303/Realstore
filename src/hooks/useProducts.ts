import {useQuery, UseQueryResult} from "@tanstack/react-query";
import {Product} from "../models/Product";

const fetchProducts = async (
    category?: string,
    productId?: number,
    minPrice?: number,
    maxPrice?: number,
    minRating?: number,
): Promise<Product[]> => {
    let url = "https://fakestoreapi.com/products";

    if (productId) {
        url = `${url}/${productId}`;
        const response = await fetch(url);
        if (!response.ok) throw new Error("Error fetching product");
        const product: Product = await response.json();
        return [product]; // Return a single product wrapped in an array
    } else if (category) {
        url = `${url}/category/${category}`;
    }

    const response = await fetch(url);
    if (!response.ok) throw new Error("Error fetching products");
    let products: Product[] = await response.json();

    // Filter products based on price and rating
    if (minPrice !== undefined) {
        products = products.filter((product) => product.price >= minPrice);
    }
    if (maxPrice !== undefined) {
        products = products.filter((product) => product.price <= maxPrice);
    }
    if (minRating !== undefined) {
        products = products.filter((product) => product.rating.rate >= minRating);
    }

    return products;
};

// Custom hook to fetch products with optional filters
export const useProducts = (
    category?: string,
    productId?: number,
    minPrice?: number,
    maxPrice?: number,
    minRating?: number,
): UseQueryResult<Product[]> => {
    const queryKey = ["products", category, productId, minPrice, maxPrice, minRating];

    return useQuery<Product[]>({
        queryKey,
        queryFn: () => fetchProducts(category, productId, minPrice, maxPrice, minRating),
    });
};
