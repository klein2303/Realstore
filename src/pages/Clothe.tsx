import {useParams} from "react-router-dom";
import ActionBox from "../components/Clothe/ActionBox";
import Navbar from "../components/Navbar/Navbar";
import Footer from "../components/Footer/Footer";
import styles from "../styles/Clothe.module.css";

const Clothe = () => {
    const {id} = useParams<{id: string}>();
    const productId = id ? parseInt(id) : 0; // Ensure id is parsed correctly

    return (
        <>
            <Navbar />
            <main className={styles.page}>
                <ActionBox productId={productId} />
            </main>
            <Footer />
        </>
    );
};

export default Clothe;
