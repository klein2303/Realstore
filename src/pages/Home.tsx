import Navbar from "../components/Navbar/Navbar";
import Footer from "../components/Footer/Footer";
import styles from "../styles/Home.module.css";
import {FaArrowRight} from "react-icons/fa6";
import {Link} from "react-router-dom";
import woman from "/woman.svg";

const Home = () => {
    return (
        <>
            <Navbar />
            <main className={styles.homepage}>
                <section className={styles.homebox} role="region" aria-label="Introduction and Shop Promotion">
                    <article className={styles.informationbox} role="article" aria-label="Store Information">
                        <p className={styles.descriptiontext}>Real Materials</p>
                        <p className={styles.descriptiontext}>Real Style</p>
                        <h1 className={styles.shoptext}>RealStore</h1>
                        <p className={styles.informationtext}>
                            Experience premium clothing, jewelry, and electronics made from authentic materials. Choose
                            us for timeless fashion and exceptional craftsmanship.
                        </p>
                        <Link className={styles.link} to="/Shopping">
                            <button className={styles.shoppingbutton} onClick={() => "/Shopping"}>
                                Explore <FaArrowRight className={styles.arrow} />
                            </button>
                        </Link>
                    </article>
                    <img src={woman} alt="Illustration of a woman showcasing the fashion" className={styles.image} />
                </section>
            </main>
            <Footer />
        </>
    );
};

export default Home;
