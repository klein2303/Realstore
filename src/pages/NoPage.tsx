import styles from "../styles/NoPage.module.css";

const NoPage = () => {
    return (
        <main className={styles.page} role="main" aria-label="Page not found">
            <h1 id="Error-404">404</h1>
        </main>
    );
};

export default NoPage;
