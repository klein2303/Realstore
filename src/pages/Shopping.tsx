import Navbar from "../components/Navbar/Navbar";
import Footer from "../components/Footer/Footer";
import styling from "../styles/Shopping.module.css";
import Scrolling from "../components/Scrolling/Scrolling";
import Filters from "../components/Filters/Filters";
import {useEffect, useState} from "react";
import {getFilter} from "../utils/sessionStorage";

export type Filter = {
    category: string;
    values: {
        min?: number;
        max?: number;
    };
};

const Shopping = () => {
    const [selectedFilters, setSelectedFilters] = useState<Filter>({
        category: "",
        values: {min: undefined, max: undefined},
    });
    useEffect(() => {
        if (getFilter().length > 0) {
            setSelectedFilters(getFilter()[0]);
        }
    }, []);
    return (
        <>
            <Navbar />
            <main className={styling.shoppingpage}>
                <Filters selectedFilter={selectedFilters} setSelectedFilters={setSelectedFilters} />
                <Scrolling favorite={true} cart={true} selectedFilter={selectedFilters} />
            </main>
            <Footer />
        </>
    );
};

export default Shopping;
