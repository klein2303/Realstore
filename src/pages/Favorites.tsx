import Navbar from "../components/Navbar/Navbar";
import Footer from "../components/Footer/Footer";
import styles from "../styles/Favorites.module.css";
import {Link} from "react-router-dom";
import ClothingsCards from "../components/Scrolling/ClothingsCards/ClothingsCards";
import {Product} from "../models/Product";
import {useProducts} from "../hooks/useProducts";
import {getFavorites} from "../utils/localStorage";
import {useEffect, useState} from "react";

const Favorites = () => {
    const [favorites, setFavorites] = useState<number[]>([]);
    const [favoriteProducts, setFavoriteProducts] = useState<Product[]>([]);
    const {data: products, isLoading, isError} = useProducts();

    useEffect(() => {
        if (!products) {
            return;
        }
        const favorites = getFavorites();
        setFavorites(favorites);
        const currentFavoriteProducts = products.filter((product: Product) => favorites.includes(product.id));
        setFavoriteProducts(currentFavoriteProducts);
    }, [products]);

    if (isLoading) {
        return (
            <section role="status" aria-live="polite">
                Loading...
            </section>
        );
    }
    if (isError) {
        return (
            <section role="alert" aria-live="assertive">
                Error fetching products.
            </section>
        );
    }

    return (
        <>
            <Navbar />
            <main className={styles.favoritepage} role="region" aria-label="Favorite Products">
                <div className={styles.favoritecontainer}>
                    <nav className={styles.path} aria-label="Navigation">
                        <Link className={styles.pathlink} to={"/"}>
                            Home
                        </Link>{" "}
                        / Favorites
                    </nav>
                    <p className={styles.title}>FAVORITES</p>
                    <section className={styles.favoriteitems} role="list" aria-description="List of favorite products">
                        {favorites.length === 0 ? (
                            <section className={styles.message} role="status" aria-live="polite">
                                No products available.
                            </section>
                        ) : (
                            favoriteProducts.map((product: Product) => (
                                <ClothingsCards
                                    key={product.id}
                                    id={product.id}
                                    title={product.title.substring(0, 20) + "..."}
                                    price={product.price}
                                    category={product.category.charAt(0).toUpperCase() + product.category.substring(1)}
                                    image={product.image}
                                    cart={true}
                                    favorite={true}
                                />
                            ))
                        )}
                    </section>
                </div>
            </main>
            <Footer />
        </>
    );
};

export default Favorites;
