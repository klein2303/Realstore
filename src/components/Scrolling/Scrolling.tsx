import styles from "./Scrolling.module.css";
import ClothingsCards from "./ClothingsCards/ClothingsCards.tsx";
import {useProducts} from "../../hooks/useProducts.ts";
import {Product} from "../../models/Product.ts";
import {Filter} from "../../pages/Shopping.tsx";

interface scrollingProps {
    favorite: boolean;
    cart: boolean;
    selectedFilter: Filter;
}

const Scrolling = ({favorite, cart, selectedFilter}: scrollingProps) => {
    const {
        data: products,
        isLoading,
        isError,
    } = useProducts(
        selectedFilter.category,
        undefined,
        selectedFilter.values.min,
        selectedFilter.values.max,
        undefined,
    );

    if (isLoading) {
        return (
            <section role="status" aria-live="polite" aria-label="Loading">
                Loading...
            </section>
        );
    }
    if (isError) {
        return (
            <section role="alert" aria-live="assertive" aria-label="Error fetching products">
                Error fetching products.
            </section>
        );
    }

    return (
        <section className={styles.clothesSections} role="region">
            <article className={styles.rows} role="list" aria-label="Product list">
                {products && products.length > 0 ? (
                    products.map((product: Product) => (
                        <ClothingsCards
                            key={product.id}
                            id={product.id}
                            title={product.title.substring(0, 20) + "..."}
                            price={product.price}
                            category={product.category.charAt(0).toUpperCase() + product.category.substring(1)}
                            image={product.image}
                            cart={cart}
                            favorite={favorite}
                        />
                    ))
                ) : (
                    <section
                        className={styles.message}
                        role="status"
                        aria-live="polite"
                        aria-label="No products available">
                        No products available.
                    </section>
                )}
            </article>
        </section>
    );
};

export default Scrolling;
