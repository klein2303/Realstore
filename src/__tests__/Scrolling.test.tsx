import {render, screen} from "@testing-library/react";
import {describe, it, expect, vi, beforeEach} from "vitest";
import Scrolling from "../components/Scrolling/Scrolling";
import {useProducts} from "../hooks/useProducts";
import {MemoryRouter} from "react-router-dom";
import {Filter} from "../pages/Shopping";

vi.mock("../hooks/useProducts", () => ({
    useProducts: vi.fn(),
}));

describe("Scrolling Component", () => {
    const mockFilter: Filter = {
        category: "electronics",
        values: {
            min: 10,
            max: 100,
        },
    };

    beforeEach(() => {
        vi.resetAllMocks();
    });

    it("renders the loading state", () => {
        (useProducts as jest.Mock).mockReturnValue({
            data: null,
            isLoading: true,
            isError: false,
        });

        render(
            <MemoryRouter>
                <Scrolling favorite={false} cart={false} selectedFilter={mockFilter} />
            </MemoryRouter>,
        );

        expect(screen.getByRole("status", {name: /loading/i})).toBeInTheDocument();
    });

    it("renders the error state", () => {
        (useProducts as jest.Mock).mockReturnValue({
            data: null,
            isLoading: false,
            isError: true,
        });

        render(
            <MemoryRouter>
                <Scrolling favorite={false} cart={false} selectedFilter={mockFilter} />
            </MemoryRouter>,
        );

        expect(screen.getByRole("alert", {name: /error fetching products/i})).toBeInTheDocument();
    });

    it("renders the product list when products are available", () => {
        const mockProducts = [
            {
                id: 1,
                title: "Mock Product 1",
                price: 50,
                category: "electronics",
                image: "mock-image-1.jpg",
            },
            {
                id: 2,
                title: "Mock Product 2",
                price: 75,
                category: "electronics",
                image: "mock-image-2.jpg",
            },
        ];

        (useProducts as jest.Mock).mockReturnValue({
            data: mockProducts,
            isLoading: false,
            isError: false,
        });

        render(
            <MemoryRouter>
                <Scrolling favorite={false} cart={false} selectedFilter={mockFilter} />
            </MemoryRouter>,
        );

        const productList = screen.getByRole("list", {name: /product list/i});
        expect(productList).toBeInTheDocument();
        expect(screen.getAllByRole("region", {name: /product card for/i})).toHaveLength(2);
    });

    it("renders the 'No products available' message when the product list is empty", () => {
        (useProducts as jest.Mock).mockReturnValue({
            data: [],
            isLoading: false,
            isError: false,
        });

        render(
            <MemoryRouter>
                <Scrolling favorite={false} cart={false} selectedFilter={mockFilter} />
            </MemoryRouter>,
        );

        const noProductsMessage = screen.getByRole("status", {name: /no products available/i});
        expect(noProductsMessage).toBeInTheDocument();
    });

    it("matches the snapshot when products are available", () => {
        const mockProducts = [
            {
                id: 1,
                title: "Mock Product 1",
                price: 50,
                category: "electronics",
                image: "mock-image-1.jpg",
            },
        ];

        (useProducts as jest.Mock).mockReturnValue({
            data: mockProducts,
            isLoading: false,
            isError: false,
        });

        const {container} = render(
            <MemoryRouter>
                <Scrolling favorite={false} cart={false} selectedFilter={mockFilter} />
            </MemoryRouter>,
        );

        expect(container).toMatchSnapshot();
    });
});
