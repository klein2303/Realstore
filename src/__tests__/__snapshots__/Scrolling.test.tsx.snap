// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Scrolling Component > matches the snapshot when products are available 1`] = `
<div>
  <section
    class="_clothesSections_d7aa43"
    role="region"
  >
    <article
      aria-label="Product list"
      class="_rows_d7aa43"
      role="list"
    >
      <article
        aria-label="Product card for Mock Product 1..."
        class="_clothingSection_47ab48"
        role="region"
      >
        <figure
          class="_imgContainer_47ab48"
        >
          <a
            href="/Clothe/1"
          >
            <img
              alt="Image of Mock Product 1..."
              class="_clothingImage_47ab48"
              src="mock-image-1.jpg"
            />
          </a>
        </figure>
        <section
          class="_descriptionSpace_47ab48"
        >
          <header
            class="_productType_47ab48"
          >
            <h5
              class="_category_47ab48"
            >
              Electronics
            </h5>
            <section
              aria-label="Product Actions"
              class="_IconSpace_47ab48"
            />
          </header>
          <footer
            class="_nameAndPrice_47ab48"
          >
            <span
              class="_productname_47ab48"
            >
              <h4
                class="_h4text_47ab48"
              >
                Mock Product 1...
              </h4>
            </span>
            <span
              class="_price_47ab48"
            >
              <h4
                class="_h4text_47ab48"
              >
                $ 
                50
              </h4>
            </span>
          </footer>
        </section>
      </article>
    </article>
  </section>
</div>
`;
